File: LookAndSay/LookAndSayFSharp/Module1.fs
namespace LookAndSayFSharp

module Module1 =
    type Group = {digit : char; count : int}

    let lookAndSay (str:string) =
        let stringDigits = str |> Seq.toList

        let incrementCount group = {group with count = group.count + 1};

        let rec getOneGroup nonEmptyDigits = 
            match nonEmptyDigits with
                | x :: y :: rest when x = y -> getOneGroup(y :: rest) |> incrementCount
                | x :: _ -> { digit = x; count = 1}

        let getOneGroupAndRemainingDigits digits =
            match digits with
                | [] -> None
                | _ ->
                    let group = getOneGroup digits
                    Some(group, digits |> List.skip group.count)

        let encodeGroup group = group.count.ToString() + group.digit.ToString()

        stringDigits
            |> Seq.unfold getOneGroupAndRemainingDigits
            |> Seq.map encodeGroup
            |> String.concat ""

File: LookAndSay/LookAndSayFunctionalCSharp/Module1.cs
using System;
using System.Collections.Immutable;
using System.Linq;
using MoreLinq;

namespace LookAndSayFunctionalCSharp
{
    public static class Module1
    {
        record Group (char Digit, int Count)
        {
            public string Encode() => Count.ToString() +  Digit;
        }

        public static string LookAndSay(string str)
        {
            var stringDigits = str.ToImmutableArray();

            Group IncrementCount(Group group) => group with { Count = group.Count + 1 };

            Group GetOneGroup (ImmutableArray<char> nonEmptyDigits) =>
                nonEmptyDigits switch
                {
                    { Length: >= 2} when nonEmptyDigits[0] == nonEmptyDigits[1] =>
                        IncrementCount(
                            GetOneGroup(nonEmptyDigits.Skip(1).ToImmutableArray())),
                    _ => new Group(nonEmptyDigits[0], 1)
                };
            
            (Group group, ImmutableArray<char> remainingDigits)? GetOneGroupAndRemainingDigits(ImmutableArray<char> digits)
            {
                if (digits.IsEmpty)
                    return null;

                var group = GetOneGroup(digits);

                return (group, digits.Skip(group.Count).ToImmutableArray());
            }

            var groups = MoreEnumerable.Unfold(
                state: stringDigits,
                generator: GetOneGroupAndRemainingDigits,
                predicate: x => x.HasValue,
                stateSelector: x => x.Value.remainingDigits,
                resultSelector: x => x.Value.group);

            return string.Concat(groups.Select(x => x.Encode()));
        }
    }
}
